mmdet/models/detectors/dino.py

```diff
@@ -0,0 +1,535 @@
+        inter_states, references = self.decoder(
+            query=query,
+            value=memory,
+            key_padding_mask=memory_mask,
+            self_attn_masks=dn_mask,
+            reference_points=reference_points,
+            spatial_shapes=spatial_shapes,
+            level_start_index=level_start_index,
+            valid_ratios=valid_ratios,
+            reg_branches=self.bbox_head.reg_branches)
+
+        inter_states = inter_states.permute(0, 2, 1, 3)
+
+        # if dn_label_query is not None and dn_label_query.size(1) == 0:
+        #     # NOTE: If there is no target in the image, the parameters of
+        #     # label_embedding won't be used in producing loss, which raises
+        #     # RuntimeError when using distributed mode.
+        #     inter_states[0] += self.label_embedding.weight[0, 0] * 0.0
+
+        decoder_outputs_dict = dict(
+            hidden_states=inter_states, references=references)
+        return decoder_outputs_dict
+
+
+class CdnQueryGenerator:
+    # TODO: Should this encapsulation be broken?
+
+    def __init__(self,
+                 num_query,
+                 hidden_dim,
+                 num_classes,
+                 noise_scale=dict(label=0.5, box=0.4),
+                 group_cfg=dict(
+                     dynamic=True, num_groups=None, num_dn_queries=None)):
+        self.num_query = num_query
+        self.hidden_dim = hidden_dim
+        self.num_classes = num_classes
+        self.label_noise_scale = noise_scale['label']
+        self.box_noise_scale = noise_scale['box']
+        self.dynamic_dn_groups = group_cfg.get('dynamic', False)
+        if self.dynamic_dn_groups:
+            assert 'num_dn_queries' in group_cfg, \
+                'num_dn_queries should be set when using ' \
+                'dynamic dn groups'
+            self.num_dn = group_cfg['num_dn_queries']
+        else:
+            assert 'num_groups' in group_cfg, \
+                'num_groups should be set when using ' \
+                'static dn groups'
+            self.num_dn = group_cfg['num_groups']
+        assert isinstance(self.num_dn, int) and self.num_dn >= 1, \
+            f'Expected the num in group_cfg to have type int. ' \
+            f'Found {type(self.num_dn)} '
+
+    def get_num_groups(self, group_queries=None):
+        """
+        Args:
+            group_queries (int): Number of dn queries in one group.
+        """
+        if self.dynamic_dn_groups:
+            assert group_queries is not None, \
+                'group_queries should be provided when using ' \
+                'dynamic dn groups'
+            if group_queries == 0:
+                num_groups = 1
+            else:
+                num_groups = self.num_dn // group_queries
+        else:
+            num_groups = self.num_dn
+        if num_groups < 1:
+            num_groups = 1
+        return int(num_groups)
+
+    def __call__(self, batch_data_samples, label_enc=None):
```

Comment:
Why not make it as a nn.module and initialize label_enc in this module?

Reply:
Haha, actually I also had the thought recently.